Change the assert check for kdfrounds and bump default rounds to 2^7

nevun opened the pull request:
The assert before checked if kdfrounds was 0 or 42 but people might want to bump it
and still support older formats.

py-bcrypt warns for anything below 50 so bumping to 128 also gets rid of a runtime warning:

/usr/local/lib/python2.7/dist-packages/bcrypt-3.1.3-py2.7-linux-x86_64.egg/bcrypt/__init__.py:137: UserWarning: Warning: bcrypt.kdf() called with only 42 round(s). This few is not secure: the parameter is linear, like PBKDF2.

bjornedstrom commented:
Thanks for the patch. 42 is hardcoded in the reference OpenBSD signify implementation (at least last ime I checked :-) ) and is used here for compatibility.

If you have time I think it'd be cool to have the option to configure this without breaking portability. Suggestion?

nevun commented:
Ok, valid point! I looked at the reference implementation and it looks like everything is done right, the rounds are embedded in the key and actually used. Just as in your implementation, Things just seems like they should work.

I tested by changing the reference impl to using 128, generated keys and signed a message and verified it. Then I changed the default back to 42 and made sure I could still verify it with the system signify. I then took those keys and signed message and added a unittest that I will push to this branch.

step 1: compiled a signify with 128 rounds default, generated key.sec, key.pub and message.sig

step 2: changed the default back to 42 and tested it with system signify
```
$ cat key.pub
untrusted comment: signify public key
RWT16WVHxrWWJzuyLW1zO7uTWPPejer1uX0PX4Y/1bkwaDgnbeU6dHnL
$ cat message.sig
untrusted comment: signature from signify secret key 128 rounds
RWT16WVHxrWWJ2fhavXHp1SDGZLzI8vpNxWyvbzXYIptU/L5xzZ/Sq+yDwFJ3u8W1Xi4CmEy+0xfM59dk/yVp6MLAifaGG9obQI=
my message
$ signify -V -m message -p key.pub -e
Signature Verified
$ cat message
my message
$
```
step 3: took the system 42-rounds-signify and signed a new message using the key generated by the 128-rounds signify:

```
$ cat > mess2
yolo
$ signify -S -s key.sec -m mess2 -e
passphrase:
$ cat mess2.sig
untrusted comment: verify with key.pub
RWT16WVHxrWWJyHkAYJ5dggrhDwdnFDDWNQnvTLGjadedSJDbzFPXBf3DYOEAT6aBbm26+/GEdBJaSE7k+eFyrQM+mv62atDAgY=
yolo
$
```

Step 4: verified that it works with your python module:
```
$ python
Python 2.7.13 (default, Feb  5 2017, 15:42:03)
[GCC 4.2.1 20070719 ] on openbsd6
>>> from signify.pure import PublicKey, Signature, verify_embedded
>>> pk = PublicKey.from_bytes(open("./key.pub").read())
>>> embedded_signature = Signature.from_bytes(open("./mess2.sig").read())
>>> print(verify_embedded(pk, embedded_signature))
yolo
```
..and added a unittest using the same keys that I used in this comment.

nevun commented:
Uhm, but maybe you just meant adding an method to the python-signify interface so one could set number of rounds.. :)

bjornedstrom commented:
Yeah I think that would be cool! Ideally the reference signify binary, and this implementation, should be compatible with each other out of the box (without having to recompile the reference signify). So removing the assert is cool, but changing the default from 42 to 2^7 is more controversial. If you can make it configurable (maybe from the scripts) that'd be sweet :)

nevun commented:
Björn Edström <notifications@github.com> writes:

> Yeah I think that would be cool! Ideally the reference signify binary, and this implementation, should be compatible with each other out of the box (without having to recompile the reference signify). So removing the assert is cool, but changing the default from 42 to 2^7 is more controversial. If you can make it configurable (maybe from the scripts) that'd be sweet :)
> 

Yeah, or maybe see if tedu want to bump the rounds in the reference implementation.

I think I was not clear on one thing though, signify can read and sign
keys with larger rounds, I tested that as well. 

Anyway.. I will look at making it configurable in python-signify,
maybe not right now but soonish. 

Thanks for the feedback :)

nevun commented:
So.. trying to to convince the source did not help, hehe

https://marc.info/?l=openbsd-tech&m=148873588331602&w=2

I also got a nice tip from theo to just add sleep 5 so I'll just give this up for now :)
